[REVIEW] Change JIT cache location to user home

devavret opened the pull request:
~Allows jit cache to be writable by others in group.~ 
New description: https://github.com/rapidsai/cudf/pull/4857#issuecomment-611795457
<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

jlowe commented:
Doesn't this have significant data-integrity and security implications if anyone on the system can write to the JIT cache and cause other users to execute their kernels?

jrhemstad commented:
> Doesn't this have significant data-integrity and security implications if anyone on the system can write to the JIT cache and cause other users to execute their kernels?

Excellent point. Why does it need 666 permissions? Do Dask workers run as different users?

kkraus14 commented:
> > Doesn't this have significant data-integrity and security implications if anyone on the system can write to the JIT cache and cause other users to execute their kernels?
> 
> Excellent point. Why does it need 666 permissions? Do Dask workers run as different users?

No they do not

devavret commented:
> Doesn't this have significant data-integrity and security implications if anyone on the system can write to the JIT cache and cause other users to execute their kernels?

Yes, that's why in an offline discussion with @harrism , we considered a different approach: make per user directories in `/tmp`. Like `/tmp/dmakkar/cudf_0.14.0` where the cache will be stored. That way, we can keep the restrictive permission bits and also prevent different users from getting in each others' way.

I swear I removed the "Ready to review" label from this using the mobile app, just to block its merging. I don't know why it's still here.

jlowe commented:
I've wondered why the JIT cache goes to such a unique, global pathname when it has to be user-specific by default for security reasons.  If the user really wants to share the cache with other users and can live with the risks, they can explicitly setup the directory perms themselves and override the JIT cache path environment variable.  I don't think we should make that risky decision for them by default.

I think the default JIT cache path should have a user-specific, distinguishing feature (like the username somewhere in the path or somewhere under the home directory) to prevent one user from grabbing the "one true path" and preventing all other users from caching any JIT results unless they remember to set the JIT cache path environment variable to a personal location.

jlowe commented:
Sorry for the comment race, looks like we're on the same page.

> make per user directories in /tmp. Like /tmp/dmakkar/cudf_0.14.0

Personally I would avoid using a vanilla directory name in tmp like the user's login name.  It's highly likely the user has created that directory for other reasons, and may be unpleasantly surprised to find cudf silently deposited other artifacts there.  I would keep the directory name specific to cudf like `/tmp/cudf-dmakkar/0.14.0` or `/home/dmakkar/.cudf/0.14.0`

kkraus14 commented:
> I would keep the directory name specific to cudf like `/tmp/cudf-dmakkar/0.14.0` or `/home/dmakkar/.cudf/0.14.0`

+1 to doing it in the home directory

devavret commented:
> > I would keep the directory name specific to cudf like `/tmp/cudf-dmakkar/0.14.0` or `/home/dmakkar/.cudf/0.14.0`
> 
> +1 to doing it in the home directory

We'd have to make a cleanup mechanism too, right? If a user goes through using multiple versions of cudf, we'd just be consuming more and more storage from their home directory. With `/tmp`, the cleanup is generally automatic.

jlowe commented:
The problems with relying on automatic cleanup are the assumptions:
- the cleanup is there (not guaranteed)
- the cleanup will "do the right thing" (may delete files still in use, e.g.: any file older than 8 hours)

The advantage of doing our own cleanup is we can ensure it's there and can control the behavior.  If we're unwilling to clean up our own messes then I agree, /tmp is a better choice than most.  But normally application caches that persist across application invocations like the one created by cudf are kept under /home rather than /tmp.  The analogy I'm thinking of is a browser page cache which are typically kept under a browser-specific directory under /home.

kkraus14 commented:
I just checked how Numba and CuPy handle this.

Numba first tries to cache in the `__pycache__` directory alongside the Python module code which we shouldn't do. If that doesn't work it goes to `/home/user/.cache/numba`.

Cupy stores its cache in `/home/user/.cupy/kernel_cache`.

Numba's policies on cache sharing / clearing are here: https://numba.pydata.org/numba-doc/dev/developer/caching.html

I'm not finding info on any cleanup mechanisms for Cupy.

trevorsm7 commented:
I think there definitely needs to be either a version-specific key in the path or some way of automatically cleaning the cache when cudf is updated, so that stale code doesn't persist after an update.

Is it possible to put a version key inside of the cache files so that we can tell if they're stale at the time they're loaded? (Not sure if this has already been fixed, just remember it was an issue before)

devavret commented:
> The advantage of doing our own cleanup is we can ensure it's there and can control the behavior. 

>Is it possible to put a version key inside of the cache files so that we can tell if they're stale at the time they're loaded?

The thing is, we can't be sure when to clean it. Right now, caches from different versions of cudf can coexist. If a user uses different cudf versions for whatever reason, like if they're using different conda environments for different jobs (and say they might run them concurrently), we want one version to leave the different version's cache alone.

jlowe commented:
> The thing is, we can't be sure when to clean it.

How is this any different than the OS cleaner which knows even _less_ about what it is cleaning?  Is it doing something that a cleaner we develop cannot do?

devavret commented:
> How is this any different than the OS cleaner which knows even _less_ about what it is cleaning? Is it doing something that a cleaner we develop cannot do?

libcudf is pretty defensive about using jit file cache. Meaning, it tries to find the cache and if it doesn't exist, it recompiles. It doesn't error when it doesn't find the cache. So it's not really a correctness problem if the OS cleans it. The OS cleaner doesn't know anything more than what we know but it also doesn't know much less than what we know. The extra info we have is that some of these caches are from older versions but that's not information we can use in making a decision to clean it. If we define rules for cleaning, we need to maintain the cleaning code. And I'm not sure if we gain much by having that control.

BTW, I'm still scoping using home dir. I'm not completely against the idea.

devavret commented:
After an online discussion, I've reverted the permission change and changed the default jit cache directory to `$HOME/.cudf/<cudf version>`. It needs to be manually cleaned for now.

kkraus14 commented on cpp/src/jit/cache.cpp:
```diff
@@ -21,19 +21,28 @@
+  }
+}
+
+// Default `LIBCUDF_KERNEL_CACHE_PATH` to `$HOME/.cudf/$CUDF_VERSION`.
 // This definition can be overridden at compile time by specifying a
 // `-DLIBCUDF_KERNEL_CACHE_PATH=/kernel/cache/path` CMake argument.
 // Use `boost::filesystem` for cross-platform path resolution and dir
 // creation. This path is used in the `getCacheDir()` function below.
```
How hard would it be to make this a runtime environment variable instead of a compile time variable? For a user needing this not in their home directory that means they'd need to recompile libcudf which can often be a non-starter.

devavret replied:
It already supports that. 

https://github.com/rapidsai/cudf/blob/c07e46ee9304db6cbc6b63e057eb777166980dc1/cpp/src/jit/cache.cpp#L54

The compile time variable is for when runtime variable is not available. The default home directory location is used when neither are available.

jrhemstad commented:
@beckernick @devavret Have we verified that the JIT cache works correctly with Dask now?

harrism commented:
>  It needs to be manually cleaned for now.

Did you add documentation of this need (probably in the top-level docs / readme)?
